okteto/okteto
# feat: add feature flag to enable context build

Fixes LAKE-94

# Proposed changes

**Introduction of serviceHasher and OktetoSmartBuildUsingContextEnvVar:**

- The `serviceHasher` generates hashes based on different criteria, such as the project commit or the build context. If the developer doesn't specify which one to use `serviceHasher` will use the one based on the feature flag
- The introduction of `OktetoSmartBuildUsingContextEnvVar` as a variable to control whether to use the build context for hashing. This provides flexibility in choosing between different hashing strategies.

**Changes in oktetoBuilderConfig and build.go:**

- The `GetBuildContextHash` method in `oktetoBuilderConfig` has been removed, and its functionality has been moved to the serviceHasher.
- The usage of `getBuildHashFromCommit` has been replaced with calls to `serviceHasher` methods (HashProjectCommit, HashBuildContext, and HashService). This makes that we don't have to rely on a magic string that is set inside the `getBuildHashFromCommit` function. 
- The `BuildContextHash` is now calculated using serviceHasher. When we want to specify if we want to hash using an specific method we have those methods but if we want to hash based on the feature flag we can do it also.

**Changes in tests:**

I have modified the tests in order to use the new serviceHasher instead of the previous method.

Ensure that the new hasher still hashes as the way it was before, by not changing the tests

## How to validate

1. Build a project as usual
2. Set the feature flag(`OKTETO_SMART_BUILDS_USING_BUILD_CONTEXT`) to true
3. Build again the project and check that the commit is different

## CLI Quality Reminders 🔧

For both authors and reviewers:

- Scrutinize for potential regressions
- Ensure key automated tests are in place
- Build the CLI and test using the validation steps
- Assess Developer Experience impact (log messages, performances, etc)
- If too broad, consider breaking into smaller PRs
- Adhere to our [code style](https://github.com/okteto/okteto/blob/master/docs/code-style.md) and [code review](https://github.com/okteto/okteto/blob/master/docs/code-review.md) guidelines
